### PR TITLE
Allow data model to be built without an organization root.

### PR DESCRIPTION
**Pull request by blueandgold**
This PR allows the data model to be built without an organization root, so that scanners can have the data they needed for scanning.  This is especially useful, when the root is a folder.

However, this means that Explain will not work fully & properly without the whole organization.  So after discuss with team, implementing a check to disable Explain functionalities, if we are not using the organization root.

image

@ahoying @joecheuk I will add test coverage if this looks good.  One thing I am not too thrilled about, is the code repetition on the `is_supported` checks.  We discussed not using decorators, so if you can think of anything cooler to do here, please let me know.

### CODE REVIEW

File: google/cloud/forseti/services/explain/service.py [hunk 1]
```diff
@@ -79,10 +97,16 @@ def ListResources(self, request, context):
         Returns:
             object: proto message of list of resources
         """
+        reply = explain_pb2.ListResourcesReply()
+
+        if not self.is_supported:
```
[the remark this answers is not in the document]
Comment (ahoying) [reply google/cloud/forseti/services/explain/service.py]: The method must return the exact Message type specified in the Service configuration:

https://github.com/GoogleCloudPlatform/forseti-security/blob/dev/google/cloud/forseti/services/explain/explain.proto#L19

You can't create a new message type just for this error, it has to return an empty proto of the type specified.

File: google/cloud/forseti/services/explain/service.py [hunk 2]
```diff
@@ -79,10 +97,16 @@ def ListResources(self, request, context):
...
+            context.set_code(StatusCode.FAILED_PRECONDITION)
```
Comment (ahoying) [inline google/cloud/forseti/services/explain/service.py]: You could create a small private helper method and use that.

```
def _set_not_supported_status(context, reply):
  context.set_code(..)
  context.set_details(..)
  return reply
```

Then in you can just do:

```
if not self.is_supported:
  return _set_not_supported_status(context, reply)
```

File: google/cloud/forseti/services/explain/service.py [hunk 3]
```diff
@@ -47,6 +52,18 @@ def _get_handle(self, context):
...
 
+    def _determine_is_supported(self):
+        """Determine whether Explainer is supported for clients to use.
+
+        Returns:
+            bool: True if Explainer module can be used.
+        """
+        root_resource_id = (
+            self.explainer.config.inventory_config.root_resource_id)
+        if 'organizations' in root_resource_id:
```
Comment (ahoying) [inline google/cloud/forseti/services/explain/service.py]: Is this the only thing that needs to be checked? What about if dwd is not configured?

File: google/cloud/forseti/services/explain/service.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (joecheuk) [reply google/cloud/forseti/services/explain/service.py]: Thanks for the clarifications!

File: tests/services/explain/service_test.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        self.grpc_explainer = service.GrpcExplainer(mock_explainer_api)
+
+    def tearDown(self):
+        """Tear down method."""
+        ForsetiTestCase.tearDown(self)
+
+    def test_determine_is_supported(self):
+        inventory_config = self.grpc_explainer.explainer.config.inventory_config
+        inventory_config.root_resource_id = 'organizations/11111'
+        self.assertTrue(self.grpc_explainer._determine_is_supported())
```
Comment (ahoying) [inline tests/services/explain/service_test.py]: Nit, make this two different tests so on assert failing doesn't prevent the second assert from getting run.

File: tests/services/explain/service_test.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+        inventory_config.root_resource_id = 'folders/22222'
+        self.assertFalse(self.grpc_explainer._determine_is_supported())
+
+    def test_failed_precondition(self):
```
Comment (ahoying) [inline tests/services/explain/service_test.py]: I would love to see a test like https://github.com/GoogleCloudPlatform/forseti-security/blob/dev/tests/services/api_tests/model_test.py#L134 for a functional integration test of the service when is_supported is false.

Comment (blueandgold) [reply tests/services/explain/service_test.py]: Done, good tip.

File: tests/services/explain/explain_test.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+            GrpcExplainerFactory,
+            GrpcInventoryFactory,
+            GrpcModellerFactory,
+        ])
+
+
+class ExplainerTest(ForsetiTestCase):
+    """Test based on declarative model."""
+
+    def setUp(self):
```
Comment (ahoying) [inline tests/services/explain/explain_test.py]: Nit: No need to have setUp here if it isn't used.

File: tests/services/explain/explain_test.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        pass
+
+    def test_explain_is_supported(self):
+        """Test explain is supported."""
+
+        inventory_config_with_organization_root = (
+            InventoryConfig(gcp_api_mocks.ORGANIZATION_ID, '', {}, '')                                                   )
+        self.setup = create_tester(
```
Comment (ahoying) [inline tests/services/explain/explain_test.py]: Nit: you don't need to assign this variable to self, you could just create a local variable instead (same below).

File: tests/services/explain/explain_test.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        """Test explain is not supported."""
+
+        inventory_config_with_folder_root = (
+            InventoryConfig(gcp_api_mocks.FOLDER_ID, '', {}, '')                                                   )
+        self.setup = create_tester(inventory_config_with_folder_root)
+
+        def test(client):
+            """Test implementation with API client."""
+            try:
+                client.explain.list_roles('')
```
Comment (ahoying) [inline tests/services/explain/explain_test.py]: Nit: you could use self.assertRaisesRegex here instead. https://docs.python.org/2/library/unittest.html#unittest.TestCase.assertRaisesRegexp.

File: tests/services/explain/explain_test.py (hunk 6 above)
Comment (blueandgold) [reply tests/services/explain/explain_test.py]: Done.

File: tests/services/explain/explain_test.py (hunk 7 above)
Comment (blueandgold) [reply tests/services/explain/explain_test.py]: Done

File: tests/services/explain/explain_test.py (hunk 8 above)
Comment (blueandgold) [reply tests/services/explain/explain_test.py]: Great tip, didn't know about this before.

**Pull request merged**